### PR TITLE
[huggingface_pytorch] Training - update for Transformers to 4.28.1 PyTorch 2.0

### PR DESCRIPTION
**Pull request by JingyaHuang**
*GitHub Issue #2986  :*

### Description

This PR updates Hugginface's PyTorch DLC for inference. Here are the corresponding updated dependencies versions:

* `transformers: 4.28.1`
* `datasets: 2.12.0`
* `evaluate: 0.4.0`
* `accelerate: 0.19.0`
* `torch: 2.0.0`
* `diffusers: 0.16.1`

And extra tests:

- [x] PyTorch 2.0 compile test
- [x] Diffusers test

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    url = (
+        f"https://raw.githubusercontent.com/huggingface/diffusers/{branch}/examples/{task}/{script}"
+    )
+    os.system(f"wget {url}")
+
+    # create default accelerate config
+    os.system("accelerate config default")
+
+    script_path = git_config["script"]
```
Comment (saimidu) [inline test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py:41]: Where does `git_config` come from?

**Review by saimidu — COMMENTED** [review]
Should there be another test added for the `evaluate` feature?

File: test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py (hunk 1 above)
Comment (JingyaHuang) [reply test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py]: just removed this line.

File: test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    parser.add_argument("--dataset_name", type=str, default="huggan/flowers-102-categories")
+    parser.add_argument("--resolution", type=int, default=64)
+    parser.add_argument("--output_dir", type=str, default="/opt/ml/model")
+    parser.add_argument("--train_batch_size", type=int, default=4)
+    parser.add_argument("--num_epochs", type=int, default=1)
+    parser.add_argument("--gradient_accumulation_steps", type=int, default=1)
+
+    args = parser.parse_args()
+
+    diffusers_version = "0.16.1"
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply test/sagemaker_tests/huggingface_pytorch/training/resources/scripts/diffusers_entrypoint.py]: Applied what you suggested, thanks!

File: huggingface/pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu [hunk 3]
```diff
@@ -10,12 +10,21 @@ LABEL dlc_major_version="1"
...
 
 # install Hugging Face libraries and its dependencies
 RUN pip install --no-cache-dir \
 	transformers[sklearn,sentencepiece,audio,vision]==${TRANSFORMERS_VERSION} \
-	datasets==${DATASETS_VERSION}
+	datasets==${DATASETS_VERSION} \
+	diffusers==${DIFFUSERS_VERSION} \
+	Jinja2 \
+	tensorboard \
+	"protobuf>=3.20.3,<4" \
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply huggingface/pytorch/training/docker/2.0/py3/cu118/Dockerfile.gpu]: seems that transformers 4.29 pinged protobuf to an old version. I removed the version constraints, 3.19.3 shall be installed but it should be fine.

**Comment by JingyaHuang** [comment]
> Should there be another test added for the `evaluate` feature?

Yep, I replaced previous sklean metrics by those in evaluate.

**Comment by saimidu** [comment]
It looks like the diffuser test never ran because of:
```
TypeError: _test_diffusers_model() missing 1 required positional argument: 'instance_type'
```

All other sagemaker integration tests for the image failed due to:
```
AttributeError: 'TrainingArguments' object has no attribute 'distributed_state'
```

**Comment by JingyaHuang** [comment]
> It looks like the diffuser test never ran because of:
> 
> ```
> TypeError: _test_diffusers_model() missing 1 required positional argument: 'instance_type'
> ```
> 
> All other sagemaker integration tests for the image failed due to:
> 
> ```
> AttributeError: 'TrainingArguments' object has no attribute 'distributed_state'
> ```

Hi @saimidu,

The first issue is fixed. For the second issue, sagemaker distributed training features seem broken since the transformers 4.29.*released this week, let's continue with 4.28.*. I will discuss this break internally with Transformers team. 

And could you help me with the sanity test? Thanks!

**Comment by JingyaHuang** [comment]
Hi @saimidu,

I just pinged the tensorflow to 2.12.0, TF is needed only as a dependency for `torch.utils.tensorboard`, so we can update it as far as it works with the tensorboard feature, let's see if the CI is green.

For the `os_scan_allowlist`, it is usually DLC team who takes care of it, maybe @arjkesh could help?

**Comment by saimidu** [comment]
@JingyaHuang All tests seem to pass with tensorflow removed, as per CI on commit [123450d](https://github.com/aws/deep-learning-containers/pull/2993/commits/123450d0822acbff7129aebf804ae10c1825e239).

Removing tensorflow from ecr-enhanced-scan allowlist and enabling other SM integration tests for the image to move forward.

**Review by saimidu — APPROVED** [review]
Approved, unless any further changes are necessary. @JingyaHuang Shall we merge this PR?

**Review by saimidu — DISMISSED** [review]
PR approved for merge.

**Review by saimidu — APPROVED** [review]

**Pull request merged**